File: apps/agentshield/internal/server/runtime_check_activity.go
package server

import (
	"errors"
	"net/http"
	"strings"

	"siq-agent-security/apps/agentshield/internal/receipt"
	"siq-agent-security/apps/agentshield/internal/runtimecheck"
)

type runtimeCheckActivity struct {
	Schema     string           `json:"schema_version"`
	CheckID    string           `json:"check_id"`
	InstanceID string           `json:"instance_id"`
	Snapshot   string           `json:"snapshot"`
	Prefix     bool             `json:"prefix_valid"`
	History    string           `json:"history_integrity"`
	Activity   taskActivityItem `json:"activity"`
}

// Resolve against the verified full snapshot, never a paginated receipt list
// or an inferred task name. Every recorded receipt must share one binding.
func resolveRuntimeCheckActivity(result runtimecheck.Result, all []receipt.Receipt, projection receipt.TaskActivities) (runtimeCheckActivity, error) {
	fail := errors.New("runtime_check_activity_binding_changed")
	if !projection.Verification.PrefixValid || projection.Verification.HistoryIntegrity == "failed" || len(result.ReceiptIDs) == 0 {
		return runtimeCheckActivity{}, fail
	}
	wanted := make(map[string]bool, len(result.ReceiptIDs))
	for _, id := range result.ReceiptIDs {
		if id == "" || wanted[id] {
			return runtimeCheckActivity{}, fail
		}
		wanted[id] = true
	}
	counts := make(map[string]int, len(wanted))
	for _, rc := range all {
		if wanted[rc.ReceiptID] {
			counts[rc.ReceiptID]++
		}
	}
	for id := range wanted {
		if counts[id] != 1 {
			return runtimeCheckActivity{}, fail
		}
	}
	for index, task := range projection.Tasks {
		matched := 0
		for _, i := range task.ReceiptIndexes {
			if wanted[all[i].ReceiptID] {
				matched++
			}
		}
		if matched != len(wanted) || task.Key.Platform != "hermes" {
			continue
		}
		page := projectActivityPage(all, projection, "tasks", index, 1)
		return runtimeCheckActivity{Schema: "local-runtime-check-activity/v1", CheckID: result.ID, InstanceID: result.InstanceID,
			Snapshot: page.Snapshot, Prefix: true, History: page.History, Activity: page.Items[0]}, nil
	}
	return runtimeCheckActivity{}, fail
}

func (s *Server) runtimeCheckActivity(w http.ResponseWriter, r *http.Request, id string) {
	if r.Method != http.MethodGet {
		w.WriteHeader(http.StatusMethodNotAllowed)
		return
	}
	if r.URL.RawQuery != "" || r.URL.ForceQuery {
		runtimeCheckError(w, errors.New("runtime_check_invalid_request"))
		return
	}
	result, err := s.runtimeChecks.Get(id)
	if err != nil {
		runtimeCheckError(w, err)
		return
	}
	if result.Status == "preparing" || result.Status == "waiting_host" || result.Status == "running" {
		runtimeCheckError(w, errors.New("runtime_check_activity_not_ready"))
		return
	}
	if len(result.ReceiptIDs) == 0 {
		runtimeCheckError(w, runtimecheck.ErrNotFound)
		return
	}
	all, projection, err := s.d.Engine.TaskActivitySnapshot()
	if err != nil {
		runtimeCheckError(w, errors.New("runtime_check_activity_unavailable"))
		return
	}
	activity, err := resolveRuntimeCheckActivity(result, all, projection)
	if err != nil {
		runtimeCheckError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, activity)
}

func runtimeCheckActivityID(path string) string {
	return strings.TrimSuffix(strings.TrimPrefix(path, "/v1/runtime-checks/"), "/activity")
}

File: apps/web/src/local/runtimeCheckActivity.ts
import { isTaskActivityPage, type TaskActivityItem, type TaskActivityPage } from './taskActivities';
import type { RuntimeCheckResult } from './types';

export interface RuntimeCheckActivity {
  schema_version: 'local-runtime-check-activity/v1';
  check_id: string;
  instance_id: string;
  snapshot: string;
  prefix_valid: true;
  history_integrity: 'verified' | 'unknown';
  activity: TaskActivityItem;
}

export function isRuntimeCheckActivity(value: unknown, result: RuntimeCheckResult): value is RuntimeCheckActivity {
  if (!value || typeof value !== 'object' || Array.isArray(value)) return false;
  const data = value as RuntimeCheckActivity;
  if (data.schema_version !== 'local-runtime-check-activity/v1' || data.check_id !== result.check_id
    || data.instance_id !== result.instance_id || !['verified', 'unknown'].includes(data.history_integrity)
    || result.receipt_ids.length === 0) return false;
  const page: TaskActivityPage = {
    schema_version: 'local-task-activities/v1', snapshot: data.snapshot, view: 'tasks', offset: 0,
    total: 1, next_offset: null, prefix_valid: data.prefix_valid, history_integrity: data.history_integrity,
    evidence_freshness: 'unknown', items: [data.activity],
  };
  return isTaskActivityPage(page, 'tasks', 0) && data.activity.binding?.platform === 'hermes'
    && data.activity.receipt_count >= result.receipt_ids.length;
}

export function runtimeCheckActivityURL(data: RuntimeCheckActivity): string {
  const binding = data.activity.binding!;
  const params = new URLSearchParams({ view: 'tasks', snapshot: data.snapshot, platform: binding.platform,
    agent_id: binding.agent_id, session_id: binding.session_id, task_id: binding.task_id });
  return `/activities/${encodeURIComponent(data.activity.activity_id)}?${params}`;
}
